### PR TITLE
WIP: [FEA] support json to struct function 

### PR DESCRIPTION
**Pull request by cindyyuanjiang**
We add support for `from_json` to a StructType.

For instance,
```
scala> val initial_df = Seq((1, """{"Zipcode":704,"ZipCodeType":"STANDARD","City":"PARC PARQUE","State":"PR"}"""))
initial_df: Seq[(Int, String)] = List((1,{"Zipcode":704,"ZipCodeType":"STANDARD","City":"PARC PARQUE","State":"PR"}))

scala> val df = initial_df.toDF("id", "value")
df: org.apache.spark.sql.DataFrame = [id: int, value: string]

scala> df.show(false)
+---+--------------------------------------------------------------------------+
|id |value                                                                     |
+---+--------------------------------------------------------------------------+
|1  |{"Zipcode":704,"ZipCodeType":"STANDARD","City":"PARC PARQUE","State":"PR"}|
+---+--------------------------------------------------------------------------+


scala> import org.apache.spark.sql.types.{StringType, StructType}
import org.apache.spark.sql.types.{StringType, StructType}

scala> val schema = new StructType().add("Zipcode", StringType, true).add("State", StringType, true)
schema: org.apache.spark.sql.types.StructType = StructType(StructField(Zipcode,StringType,true), StructField(State,StringType,true))

scala> val df_struct = df.withColumn("value",from_json(col("value"),schema))
df_struct: org.apache.spark.sql.DataFrame = [id: int, value: struct<Zipcode: string, State: string>]

scala> df_struct.show(false)
+---+---------+
|id |value    |
+---+---------+
|1  |{704, PR}|
+---+---------+
```

### CODE REVIEW

**Comment by cindyyuanjiang** [comment]
build

File: integration_tests/src/main/python/json_test.py [hunk 1]
```diff
@@ -368,3 +368,16 @@ def test_from_json_map():
         lambda spark : unary_op_df(spark, json_string_gen) \
             .select(f.from_json(f.col('a'), 'MAP<STRING,STRING>')),
         conf={"spark.rapids.sql.expression.JsonToStructs": "true"})
+
+@pytest.mark.parametrize('data_gen', [StringGen(r'{"a": "[0-9]{0,5}", "b": "[A-Z]{0,5}", "c": 1234}')])
+@pytest.mark.parametrize('schema', [StructType([StructField("a", StringType())]),
+                                    StructType([StructField("d", StringType())]),
+                                    StructType([StructField("a", StringType()), StructField("b", StringType())]),
+                                    StructType([StructField("c", LongType()), StructField("a", StringType())]),
+                                    StructType([StructField("a", StringType()), StructField("a", StringType())])
```
Comment (revans2) [inline integration_tests/src/main/python/json_test.py:377]: I only see tests for String, Long, and Struct. If we say that we support the other types we really should have tests for them. This needs to include things like STRUCT of STRUCTs and STURCTs of LISTS.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 2]
```diff
@@ -3373,14 +3373,15 @@ object GpuOverrides extends Logging {
     expr[JsonToStructs](
       "Returns a struct value with the given `jsonStr` and `schema`",
       ExprChecks.projectOnly(
-        TypeSig.MAP.nested(TypeSig.STRING),
+        (TypeSig.STRUCT + TypeSig.MAP).nested(TypeSig.all),
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3376]: We need a way so that MAP type is only supported if it is a MAP<STRING, STRING> and only if it is at the top level. Some of this can be done with a change to this line. But we need more than this and ideally have some tests to verify that we do fall back properly.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 3]
```diff
@@ -3373,14 +3373,15 @@ object GpuOverrides extends Logging {
...
         (TypeSig.STRUCT + TypeSig.MAP + TypeSig.ARRAY).nested(TypeSig.all),
         Seq(ParamCheck("jsonStr", TypeSig.STRING, TypeSig.STRING))),
       (a, conf, p, r) => new UnaryExprMeta[JsonToStructs](a, conf, p, r) {
         override def tagExprForGpu(): Unit =
           GpuJsonScan.tagJsonToStructsSupport(a.options, this)
 
         override def convertToGpu(child: Expression): GpuExpression =
+          // GPU implementation currently does not support duplicated json key names in input
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3384]: Does the CPU? If so what happens when they show up? Should we fall back to the CPU if we see it?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 4]
```diff
@@ -30,8 +34,118 @@ case class GpuJsonToStructs(
...
+            withResource(cudf.Scalar.fromInt(0)) { zero =>
+              lengths.lessOrEqualTo(zero)
+            }
+          }
+          withResource(isEmpty) { isEmpty =>
+            isNull.binaryOp(cudf.BinaryOp.NULL_LOGICAL_OR, isEmpty, cudf.DType.BOOL8)
+          }
+        }
+        closeOnExcept(isNullOrEmptyInput) { _ =>
+          withResource(isNullOrEmptyInput.ifElse(emptyRow, stripped)) { cleaned =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:55]: So CUDF cannot deal with leading or training white space in JSON lines???

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 5]
```diff
@@ -30,8 +34,118 @@ case class GpuJsonToStructs(
...
+            combined.copyToHost()
+          }
+          // Step 4: Have cudf parse the JSON data
+          val (names, rawTable) = withResource(combinedHost) { combinedHost =>
+            val data = combinedHost.getData
+            val start = combinedHost.getStartListOffset(0)
+            val end = combinedHost.getEndListOffset(0)
+            val length = end - start
+
+            withResource(cudf.Table.readJSON(cudf.JSONOptions.DEFAULT, data, start,
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:105]: This is having CUDF do name and type inference. Is that really what we want? Should we do this like we do for regular JSON parsing? (Never mind turns out we do the same thing in the JSON reader??? What are we doing that? It is a huge waste of memory. Can we please file a follow on issue to fix it both here and in the JSON reader. Bonus points if we can combine the code reader code together.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 6]
```diff
@@ -30,8 +34,118 @@ case class GpuJsonToStructs(
...
+            // Step 6: get the data based on input struct schema
+            val columns = fieldNames.safeMap { case (name, dtype) =>
+              val i = names.indexOf(name)
+              if (i == -1) {
+                GpuColumnVector.columnVectorFromNull(numRows, dtype)
+              } else {
+                rawTable.getColumn(i).incRefCount
+              }
+            }
+
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:133]: We have to be able to cast the data to the appropriate return type. We are not doing it here at all. So if a user asks for an int, but a long shows up we will get errors. Or what about a String and a Struct shows up?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 3 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3384]: Yes, for duplicated keys in the first level, CPU will extract the value for the key that shows up the latest in the json string. I have not yet find an efficient way to identify duplicated keys. After calling `cudf.Table.readJSON`, the result table only contains the values for the first (duplicated) key in the json string and the values for the rest of keys are no longer present in the table.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 4 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:55]: This part of implementation is inherited from https://github.com/NVIDIA/spark-rapids/pull/6211. I will look into this.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 6 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:133]: Updated changes. Thank you!

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 7]
```diff
@@ -30,8 +34,118 @@ case class GpuJsonToStructs(
...
+              lengths.lessOrEqualTo(zero)
+            }
+          }
+          withResource(isEmpty) { isEmpty =>
+            isNull.binaryOp(cudf.BinaryOp.NULL_LOGICAL_OR, isEmpty, cudf.DType.BOOL8)
+          }
+        }
+        closeOnExcept(isNullOrEmptyInput) { _ =>
+          withResource(isNullOrEmptyInput.ifElse(emptyRow, stripped)) { cleaned =>
+            withResource(cudf.Scalar.fromString("\n")) { lineSep =>
```
[the remark this answers is not in the document]
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:56]: I ran some tests and it does not look like `\r` is supported. Updated changes in implementation. Thank you!

**Comment by cindyyuanjiang** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 8]
```diff
@@ -30,8 +34,128 @@ case class GpuJsonToStructs(
...
+                  s"but got ${rawTable.getRowCount}")
+            }
+
+            // Step 6: get the data based on input struct schema
+            val columns = fieldNames.safeMap { case (name, dtype) =>
+              val i = names.indexOf(name)
+              if (i == -1) {
+                GpuColumnVector.columnVectorFromNull(numRows, dtype)
+              } else {
+                rawTable.getColumn(i).castTo(GpuColumnVector.getRapidsType(dtype))
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:140]: Please use GpuCast instead of castTo. IT does a lot more checking to make sure that we parse things correctly/etc.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 9]
```diff
@@ -3373,14 +3373,24 @@ object GpuOverrides extends Logging {
...
+        TypeSig.MAP.nested(TypeSig.STRING) + TypeSig.STRUCT.nested(TypeSig.all),
         (TypeSig.STRUCT + TypeSig.MAP + TypeSig.ARRAY).nested(TypeSig.all),
         Seq(ParamCheck("jsonStr", TypeSig.STRING, TypeSig.STRING))),
       (a, conf, p, r) => new UnaryExprMeta[JsonToStructs](a, conf, p, r) {
         override def tagExprForGpu(): Unit =
+          a.schema match {
+            case map: MapType => {
+              if(!(map.keyType == StringType && map.valueType == StringType)) {
+                willNotWorkOnGpu("JsonToStructs only supports Map<String, String> for " +
+                                 "MapType schema")
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3385]: it may help the user debug the fallback if we show the actual map type<keytype,vlauetype> in the error message

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 10]
```diff
@@ -30,8 +34,128 @@ case class GpuJsonToStructs(
...
+    }
+  }
+
+  private def processFieldNames(names: Seq[(String, DataType)]): Seq[(String, DataType)] = {
+    val existingNames = Set[String]()
+    // for duplicated field names, only keep the one with the largest index
+    names.foldRight(Seq[(String, DataType)]())((elem, acc) => {
+      val (name, dtype) = elem
+      if (existingNames(name)) (null, dtype)+:acc else {existingNames += name; (name, dtype)+:acc}})
+  }
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:89]: This code is hard to follow. Can you give an example of inputs  and outputs?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 11]
```diff
@@ -30,8 +34,128 @@ case class GpuJsonToStructs(
...
+                }
+                withResource(cleaned.stringContains(returnSep)) { inputHas =>
+                  withResource(inputHas.any()) { anyReturnSep =>
+                    if (anyReturnSep.isValid && anyReturnSep.getBoolean) {
+                      throw new IllegalArgumentException("We cannot currently support parsing " +
+                          "JSON that contains a carriage return in it")
+                    }
+                  }
+                }
+              }
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:74]: not sure about performance of containsRe but maybe cheaper to create a regex program for a regex to the tune of `[\r\n]` to launch just one kernel instead https://github.com/rapidsai/cudf/blob/fd1afc8f0dc98f4409ddac30345dbe4a4ed3bb96/java/src/main/java/ai/rapids/cudf/ColumnView.java#L3376-L3399.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 8 above)
Comment (NVnavkumar) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:140]: To add to this point, should add integration tests to test any failure conditions that are a result of casting. For example, if you pass in JSON that has arbitrary large number that you specify in your schema to use IntegerType, the same overflow exception that Spark would throw should be checked for.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 9 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3385]: Thank you! Updated the error message.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 10 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:89]: Thank you! I will add more comments.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 11 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:74]: not sure if regex is more performant. I will have to investigate more on this.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 8 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala]: I updated the implementation to use `GpuCast.doCast`. I still observed incompatible CPU vs GPU behavior with arbitrarily large numbers in input JSON strings with IntegerType schema: CPU implementation returns null, and GPU implementation casts the large numbers to integers. I documented this observation in the `compatibility.md` file. I have not yet found a way to work around this while not breaking current existing tests. If it is important to be compatible, I will investigate more into this. Thanks!

**Comment by cindyyuanjiang** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 12]
```diff
@@ -30,8 +36,156 @@ case class GpuJsonToStructs(
...
+            }
+
+            // Step 6: get the data based on input struct schema
+            val columns = fieldNames.safeMap { case (name, dtype) =>
+              val i = names.indexOf(name)
+              if (i == -1) {
+                GpuColumnVector.columnVectorFromNull(numRows, dtype)
+              } else {
+                val col = rawTable.getColumn(i)
+                doCast(col, getSparkType(col), dtype, false, false, false)
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:170]: schema should match this with the same number of columns and in the same order or we have a bug.  Use the types from struct instead of this. Or if we are going to turn it into a struct, make it into a struct before the cast, and then do the cast with just struct as the type.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 13]
```diff
@@ -30,8 +36,156 @@ case class GpuJsonToStructs(
...
+      case cudf.DType.FLOAT64 => DoubleType
+      case cudf.DType.BOOL8 => BooleanType
+      case cudf.DType.STRING => StringType
+      case cudf.DType.LIST => ArrayType(getSparkType(col.getChildColumnView(0)))
+      case cudf.DType.STRUCT =>
+        val structFields = ArrayBuffer.empty[StructField]
+        (0 until col.getNumChildren).foreach { i =>
+          val child = col.getChildColumnView(i)
+          structFields += StructField("", getSparkType(child))
+        }
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:113]: I don't think we need a mutable here
```suggestion
        val structFields = (0 until col.getNumChildren).map { i =>
          val child = col.getChildColumnView(i)
          StructField("", getSparkType(child))
        }
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala [hunk 14]
```diff
@@ -30,8 +36,156 @@ case class GpuJsonToStructs(
...
+  // Process a sequence of field names. If there are duplicated field names, we only keep the field
+  // name with the largest index in the sequence, for others, replace the field names with null.
+  // Example:
+  // Input = [("a", StringType), ("b", StringType), ("a", IntegerType)]
+  // Output = [(null, StringType), ("b", StringType), ("a", IntegerType)]
+  private def processFieldNames(names: Seq[(String, DataType)]): Seq[(String, DataType)] = {
+    val existingNames = Set[String]()
+    names.foldRight(Seq[(String, DataType)]())((elem, acc) => {
+      val (name, dtype) = elem
+      if (existingNames(name)) (null, dtype)+:acc else {existingNames += name; (name, dtype)+:acc}})
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:94]: Thanks for explanation! It will be easier to read with more line breaks IMO. 

nit: If we move existingNames into fold we can get away without a mutable Set.
```suggestion
    val zero = (Set.empty[String], Seq.empty[(String, DataType)])
    val (_, res) = names.foldRight(zero) { case ((name, dtype), (existingNames, acc)) =>
      if (existingNames(name)) {
        (existingNames, (null, dtype) +: acc)
      } else {
        (existingNames + name, (name, dtype) +: acc)
      }
    }
    res
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 12 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:170]: The value `dtype` is the type from `struct` schema. Where do we want to use the type from `struct` in `doCast`? If `dtype` is a struct, do we want to first cast `col` to `DType.STRUCT` then cast it to `dtype`? `doCast` only takes in Spark `DataType`, I will need to find the equivalent Spark type for the input cudf column `col` in some way.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 14 above)
Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala]: Thank you very much! Changes updated.

**Comment by cindyyuanjiang** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala (hunk 12 above)
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:171]: Oh, sorry I got confused. You are trying to get the from type, not the to type. I think what you have is fine then. It would be good to document it a bit to say that getSparkType is only used for the from type of a cast .

Comment (cindyyuanjiang) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuJsonToStructs.scala:173]: Thank you! I added comments to document `getSparkType`.

**Comment by cindyyuanjiang** [comment]
build

**Review by gerashegalov — APPROVED** [review]
LGTM, pending other reviews

**Review by revans2 — APPROVED** [review]
